pyfr/plugins/tavg.py

```diff
@@ -358,3 +363,248 @@ def __call__(self, intg):
+            # Merge the averages and stds
+            self._merge_data(outf)
+
+            # Merge the metadata
+            self._merge_meta(outf)
+
+    def _eval_fun_exprs(self, avars):
+        # Prepare the substitution dictionary
+        subs = dict(zip(self.anames, avars.swapaxes(0, 1)))
+
+        # Evaluate the function and return
+        return np.stack([npeval(v, subs) for v in self.fexprs], axis=1)
+
+    def _eval_fun_var(self, std, avars):
+        dfexpr = []
+
+        # Evaluate the function
+        fexprs = self._eval_fun_exprs(avars)
+
+        for avi in avars.swapaxes(0, 1):
+            # Calculate step size for finite difference
+            h = self.delta_h*np.abs(avi)
+            h[np.where(h == 0)] = self.delta_h
+
+            # Calculate derivatives for functional averages
+            avi += h
+            dfexpr.append((self._eval_fun_exprs(avars) - fexprs) / h[:,None])
+            avi -= h
+
+        # Multiply by variance and take RMS value
+        fv = np.linalg.norm(np.array(dfexpr)*std.swapaxes(0, 1)[:, :, None],
+                            axis=0)
+        return fexprs, fv
+
+    def _init_tavg_merge(self):
+        self.delta_h = np.finfo(np.float64).eps**0.5
+
+        for i, (file, stats, cfg, dt) in enumerate(self.files):
+            fields = stats.get('data', 'fields').split(',')
+            cfgsect = stats.get('tavg', 'cfg-section')
+
+            # Mapping for expression to sorted list
+            aidx = [i for i, v in enumerate(fields) if v.startswith('avg-')]
+            anames = [fields[i].removeprefix('avg-') for i in aidx]
+            amap = [i for n, i in sorted(zip(anames, aidx))]
+
+            sidx = [i for i, v in enumerate(fields) if
+                    re.match(r'\bstd-(?!fun-avg-)', v)]
+            snames = [fields[i].removeprefix('std-') for i in sidx]
+            smap = [i for n, i in sorted(zip(snames, sidx))]
+
+            if i == 0:
+                self.stats, self.cfg = stats, cfg
+
+                prec = cfg.get('backend', 'precision')
+                self.dtype = np.float32() if prec == 'single' else np.float64()
+
+                self.fields = fields
+                self.cfgsect = cfgsect
+                self.region = cfg.get(cfgsect, 'region')
+                self.uuid = file['mesh_uuid']
+
+                self.idxs = [k for k in file if re.search(r'_idxs_p\d+$', k)]
+                self.dkeys = [k for k in file if
+                              re.match(r'tavg_[^\W_]+_p\d+$', k)]
+
+                self.mapping = [(amap, smap)]
+                fnames = [x.removeprefix('fun-avg-') for x in fields if
+                          x.startswith('fun-avg-')]
+                self.anames = sorted(anames)
+                self.fnames = sorted(fnames)
+
+                # Initialise std max and avg registers
+                if self.std_all:
+                    self.std_max = np.zeros(len(anames), dtype=self.dtype)
```

Comment:
No dtype here? Just cast at the end when we write to ensure accumulation is done at the highest possible precision.